User model

JuanSanchezWolox opened the pull request:
## Summary
Created model and migrations for the user's data.
Created controller for to save user's data.
User information is validated correctly before to save data.

## Known Issues
None

## Trello Card
https://trello.com/c/v1q1gq49
https://trello.com/c/r3T3JeHx

Commit:
Created model and migration for the user's data.

JuanSanchezWolox committed:
The controller for user registration was created.

JuanSanchezWolox committed:
User information is validated correctly before to save data.

jxczz2 reviewed (changes requested):
Update branch with master and resolve conflicts

JuanSanchezWolox committed:
Merge branch 'master' into user-model

jxczz2 commented on app/models/user.js:
```diff
@@ -0,0 +1,20 @@
+      tableName: 'users'
+    }
+  );
+  user.associate = () => {
```
remove unnecessary code

jxczz2 commented on app/models/user.js:
```diff
@@ -0,0 +1,20 @@
+'use strict';
+
+module.exports = (sequelize, DataTypes) => {
+  const user = sequelize.define(
```
name of the variable where you define the model must start with a capital letter, the same in the define

jxczz2 commented on app/models/user.js:
```diff
@@ -0,0 +1,20 @@
+    'user',
+    {
+      name: DataTypes.STRING,
+      last_name: DataTypes.STRING,
```
they should be in camelCase in the model

JuanSanchezWolox replied:
But in the tech-guides they said that conventions the database and columns can be in snake_case

jxczz2 commented on app/models/user.js:
```diff
@@ -0,0 +1,20 @@
+  const user = sequelize.define(
+    'user',
+    {
+      name: DataTypes.STRING,
```
You should define in the model the properties of each field if it is unique, if it is optional, if it is a primary key etc.

jxczz2 commented on app/models/user.js:
```diff
@@ -0,0 +1,20 @@
+module.exports = (sequelize, DataTypes) => {
+  const user = sequelize.define(
+    'user',
+    {
```
id field is missing in the model

jxczz2 commented on app/controllers/users/users.controller.js:
```diff
@@ -0,0 +1,23 @@
+const { encryptar } = require('./utils');
+
+exports.createUser = (req, res) => {
+  const { name, last_name, email, password } = req.body;
```
in this case you could make use of the mappers https://github.com/Wolox/tech-guides/blob/master/nodejs/docs/node-standard-and-best-practices.md#28--mappers

jxczz2 commented on app/controllers/users/users.controller.js:
```diff
@@ -0,0 +1,23 @@
+
+exports.createUser = (req, res) => {
+  const { name, last_name, email, password } = req.body;
+  encryptar(password)
```
I think that the controller is very overloaded with logic and at first glance it is difficult to understand what is being done, so it is advisable to separate it in this case to take it to a service

jxczz2 commented on app/controllers/users/utils.js:
```diff
@@ -0,0 +1,20 @@
+
+exports.encryptar = password =>
+  new Promise((resolve, reject) => {
+    bcrypt.genSalt(10, (errGenSalt, salt) => {
```
move the number 10 to an environment variable

jxczz2 commented on app/controllers/users/utils.js:
```diff
@@ -0,0 +1,20 @@
+const bcrypt = require('bcryptjs');
```
I think that the best place for this type of logic is the helpers https://github.com/Wolox/tech-guides/blob/master/nodejs/docs/node-standard-and-best-practices.md#24--helpers

jxczz2 commented on app/middlewares/users/verify_signs_up.js:
```diff
@@ -0,0 +1,25 @@
+const { badRequest_Error } = require('../../errors');
+const logger = require('../../logger');
+
+exports.verifyEmail = (req, res, next) => {
```
According to the card requirements you should validate more fields to know if they have data, in this case I recommend using the express-validator library that the project has to validate the schema fields

JuanSanchezWolox committed:
·Update model and migration user.
·Created middleware for encryp password.
·Created folder helper for encryp and decrypt.

JuanSanchezWolox committed:
Feedback 'Pull request'
·Created mapper for signs_up.
·Updated model user with camelCase

jxczz2 commented on migrations/migrations/20210609150708-create-user.js:
```diff
@@ -0,0 +1,41 @@
+        type: Sequelize.STRING,
+        require: true
+      },
+      createdAt: {
```
remember that in the database the fields must go in snake_case

jxczz2 commented on migrations/migrations/20210609150708-create-user.js:
```diff
@@ -0,0 +1,41 @@
+        allowNull: false,
+        type: Sequelize.DATE
+      },
+      updatedAt: {
```
the same here

jxczz2 commented on app/models/user.js:
```diff
@@ -0,0 +1,43 @@
+        type: DataTypes.UUID,
+        primaryKey: true,
+        defaultValue: sequelize.UUIDV1,
+        field: 'id'
```
hint, when the field is called the same in the database you don't need to add the field field

jxczz2 commented on app/controllers/users.controller.js:
```diff
@@ -0,0 +1,15 @@
+
+exports.createUser = (req, res) => {
+  const dataUser = signUp(req.body);
+  db.User.create(dataUser)
```
For example, I should have a user service that allows me to export a method to create users, https://github.com/Wolox/tech-guides/blob/master/nodejs/docs/node-standard-and-best-practices.md#22--services

remember naming conventions for files, https://github.com/Wolox/tech-guides/blob/master/nodejs/docs/node-standard-and-best-practices.md#31--files  instead of users.controller.js put it users.js

jxczz2 commented on app/helpers/utils.js:
```diff
@@ -0,0 +1,21 @@
+
+exports.encryptar = password =>
+  new Promise((resolve, reject) => {
+    bcrypt.genSalt(parseInt(config.common.roundsBcrypt), (errGenSalt, salt) => {
```
You could check the genSaltSyn method

jxczz2 commented on app/helpers/utils.js:
```diff
@@ -0,0 +1,21 @@
+  new Promise((resolve, reject) => {
+    bcrypt.genSalt(parseInt(config.common.roundsBcrypt), (errGenSalt, salt) => {
+      if (errGenSalt) reject(errGenSalt);
+      bcrypt.hash(password, salt, (errHash, hash) => {
```
you could check the library's hashSync method

jxczz2 commented on app/helpers/utils.js:
```diff
@@ -0,0 +1,21 @@
+    });
+  });
+
+exports.comparePassword = (password, hashPassword) =>
```
Remember to follow the card requirements this is fine, but for the moment it is not necessary for the card because it is not yet the signIn part. 

also for this the library has a synchronous method compareSync

jxczz2 commented on app/middlewares/users/verify_signs_up.js:
```diff
@@ -0,0 +1,40 @@
+const { encryptar } = require('../../helpers/utils');
+
+exports.verifyEmail = (req, res, next) => {
+  const { email } = req.body;
```
The other fields still need to be validated remember that you can validate the schema with express-validator

JuanSanchezWolox committed:
·Validated params new user with express-validator.
·Created user service.

JuanSanchezWolox committed:
·Updated validation signs up.
·Created serializer signUp.

JuanSanchezWolox committed:
·Updated Serializer SignUp.

JuanSanchezWolox committed:
·Created file constants.
·Created schemes folder validators in middlewares.
